flink-runtime/src/main/java/org/apache/flink/runtime/state/CheckpointStorage.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.state;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+
+/**
+ * CheckpointStorage implements the durable storage of checkpoint data and metadata streams.
+ * An individual checkpoint or savepoint is stored to a {@link CheckpointStorageLocation},
+ * created by this class.
+ */
+public interface CheckpointStorage {
```

Comment:
I wonder if we can find a name that makes the role of this class w.r.t `CheckpointStorageLocation` clearer, in particular that this is what assigns the storage locations to the individual checkpoints?

Reply:
Another point about naming that I found confusing: Backends have a methods to create `CheckpointStorage` and `CheckpointStreamFactory`. From the names, it sounds like the storage could also do the job of the stream factories. In fact, the storage is dealing more with the metadata aspect on the checkpoint coordinator (which we call `Checkpoint`), and the factories are giving as streams that are used in checkpointing (so far, so good), but to write the snapshots on the task managers.
Maybe the problem is rooted in that checkpoint is some metadata class for the JM, as well as some concept - and in the method current names both get mixed up.

Reply:
The storage will create the stream factories later, that is the plan. Simply did not want to push that in addition into the pull request. This of this pull request as the "JobManager" side of things, the "TaskManager" side of thing is to come next.